### PR TITLE
expression: prepare `grouping` function for `roll up`

### LINKED ISSUE #42463: Prepare `grouping` function for `roll up`
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #42463

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by Defined2014** [comment]
Please change `TestShowBuiltin` test

File: expression/builtin_grouping.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+	return b.grouping(groupingID), false, nil
+}
+
+func (b *builtinGroupingSig) groupingVec(groupingIds *chunk.Column, rowNum int, result *chunk.Column) {
+	result.ResizeInt64(rowNum, false)
+	result.MergeNulls(groupingIds)
+	resContainer := result.Int64s()
+	for i := 0; i < rowNum; i++ {
+		if result.IsNull(i) {
```
Comment (AilinKid) [inline expression/builtin_grouping.go:168]: what kind of the cases like when grouping id is null? (outer join?)
or we can say, this is invalid? 

expand OP co-existed with Aggregate, and groupingID is produced in expand OP, through shuffle OP if any, and consumed by Aggregate and Upper Projection. 

Any subquery（leading join rewrite later） in aggregate args and projection list won't fill the groupingID as null

File: expression/builtin_grouping.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func (b *builtinGroupingSig) groupingImplV2(groupingID int64, metaGroupingID int64) int64 {
+	if groupingID > metaGroupingID {
+		return 1
+	}
+	return 0
+}
+
+func (b *builtinGroupingSig) groupingImplV3(groupingID int64) int64 {
```
Comment (AilinKid) [inline expression/builtin_grouping.go:127]:
```
| year | country | product    | profit | grp_year | grp_country | grp_product |
| 2001 | NULL    | NULL       |   3010 |        0 |           1 |           1 |
```
for a case like the above, the grouping(product) = 1 means the product col is filled as null for a bigger group granularity of `a`. And the actual group column is a, so we pass the corresponding grouping_id as parameter as a meta (form a id list here) to grouping function here. Make sense.

for V2, it seems rarely used ( control the grouping id > some value), it's ok, not a strong requirement

File: expression/builtin_grouping.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+		// TODO
+	}
+	return args
+}
+
+func (b *builtinGroupingSig) Clone() builtinFunc {
+	newSig := &builtinGroupingSig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	newSig.version = b.version
+	newSig.groupingIDs = b.groupingIDs
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:85]: > how about deep clone？

I think it's a common `Clone` implemention that has been applied on many functions

**Comment by xzhangxian1008** [comment]
> you can link your pb pull request here.

okk, I have put it on the top field

**Comment by xzhangxian1008** [comment]
> Please fix `TestShowBuiltin` test

okk

File: expression/builtin_grouping.go (hunk 1 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:168]: > 

ok, I will delete them

File: DEPS.bzl [hunk 4]
```diff
@@ -3402,8 +3402,9 @@ def go_deps():
         name = "com_github_pingcap_tipb",
         build_file_proto_mode = "disable_global",
         importpath = "github.com/pingcap/tipb",
-        sum = "h1:CeeMOq1aHPAhXrw4eYXtQRyWOFlbfqK1+3f9Iop4IfU=",
-        version = "v0.0.0-20230310043643-5362260ee6f7",
+        replace = "github.com/pingcap/tipb",
```
Comment (windtalker) [inline DEPS.bzl:3405]: Is the `replace = xxxx` needed?

File: expression/builtin_grouping.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+var (
+	_ functionClass = &groupingFunctionClass{}
```
Comment (windtalker) [inline expression/builtin_grouping.go:27]: I think we should use another name like grpuingimpl or something here because this function is actually not the original function used by user?

File: expression/builtin_grouping.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	sig := &builtinGroupingSig{bf, 0, map[int64]struct{}{}}
+	sig.setPbCode(tipb.ScalarFuncSig_GroupingSig)
+	return sig, nil
+}
+
+type builtinGroupingSig struct {
+	baseBuiltinFunc
+
+	// TODO these are two temporary fields for tests
+	version     uint32
```
Comment (windtalker) [inline expression/builtin_grouping.go:57]: use enum here?

File: expression/builtin_grouping.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	return b.grouping(groupingID), false, nil
+}
+
+func (b *builtinGroupingSig) groupingVec(groupingIds *chunk.Column, rowNum int, result *chunk.Column) {
+	result.ResizeInt64(rowNum, false)
+	resContainer := result.Int64s()
+	for i := 0; i < rowNum; i++ {
+		resContainer[i] = b.grouping(groupingIds.GetInt64(i))
```
Comment (windtalker) [inline expression/builtin_grouping.go:167]: use swich inside loop seems not very efficient?

File: DEPS.bzl (hunk 4 above)
Comment (xzhangxian1008) [reply DEPS.bzl:3413]: > Is the `replace = xxxx` needed?

I will delete it before merge.

File: expression/builtin_grouping.go (hunk 6 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:57]: > use enum here?

okk

File: expression/builtin_grouping.go (hunk 7 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:167]: > use swich inside loop seems not very efficient?

okk, I will fix it.

File: expression/builtin_grouping.go (hunk 5 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:27]: > I think we should use another name like grpuingimpl or something here because this function is actually not the original function used by user?

okk

**Review by AilinKid — APPROVED** [review]

File: expression/builtin_grouping.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	for key := range grouping_ids {
+		metaGroupingID = key
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingSig) checkMetadata() error {
+	version := b.getMetaVersion()
+	grouping_ids := b.getMetaGroupingIDs()
+	if version < 1 || version > 3 {
```
Comment (windtalker) [inline expression/builtin_grouping.go:105]: `1` and `3` should use enum value

File: expression/builtin_grouping.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+	return b.grouping(groupingID), false, nil
+}
+
+func (b *builtinGroupingSig) groupingVec(groupingIds *chunk.Column, rowNum int, result *chunk.Column) {
+	result.ResizeInt64(rowNum, false)
+	resContainer := result.Int64s()
+	switch b.mode {
+	case tipb.GroupingMode_ModeBitAnd:
+		for i := 0; i < rowNum; i++ {
+			resContainer[i] = b.groupingImplBitAnd(groupingIds.GetInt64(i), b.getMetaGroupingID())
```
Comment (windtalker) [inline expression/builtin_grouping.go:169]: maybe getMetaGroupingID outside the loop is better.

File: parser/ast/functions.go [hunk 10]
```diff
@@ -300,6 +300,7 @@ const (
 	TiDBShard       = "tidb_shard"
 	GetLock         = "get_lock"
 	ReleaseLock     = "release_lock"
+	Grouping        = "grouping"
```
Comment (windtalker) [inline parser/ast/functions.go:303]: shoul not add this since we actually not support it yet.

File: expression/builtin_regexp_util.go [hunk 11]
```diff
@@ -72,6 +72,13 @@ func fillNullStringIntoResult(result *chunk.Column, num int) {
 	}
 }
 
+func fillNullBytesIntoResult(result *chunk.Column, num int) {
```
Comment (windtalker) [inline expression/builtin_regexp_util.go:75]: seems this function is not used?

File: expression/builtin.go [hunk 12]
```diff
@@ -799,6 +799,7 @@ var funcs = map[string]functionClass{
 	ast.UUIDToBin:       &uuidToBinFunctionClass{baseFunctionClass{ast.UUIDToBin, 1, 2}},
 	ast.BinToUUID:       &binToUUIDFunctionClass{baseFunctionClass{ast.BinToUUID, 1, 2}},
 	ast.TiDBShard:       &tidbShardFunctionClass{baseFunctionClass{ast.TiDBShard, 1, 1}},
+	ast.Grouping:        &groupingImplFunctionClass{baseFunctionClass{ast.Grouping, 1, 1}},
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply expression/builtin.go:802]: > 

okk, I will delete it from function map

[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply expression/builtin.go:802]: > I'm not sure it is right to map `ast.Grouping` to `groupingImpl` function. `ast.Grouping` is a function that used by user directly, for example, user can use `grouping(time_col)` here, but `groupingImpl` actually is a rewritten function that only accept `UInt64` as its input. I think this pr should only focus on the implementation of `groupingImplFunctionClass`, and should not change the function map, or at least not use `ast.Grouping` as the key and the function name.

Current ut needs these info to create functions, I think they are necessary in this pr so far.

File: expression/builtin_grouping.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+	if err != nil {
+		return nil, err
+	}
+	bf.tp.SetFlen(1)
+	sig := &builtinGroupingSig{bf, 0, map[int64]struct{}{}}
+	sig.setPbCode(tipb.ScalarFuncSig_GroupingSig)
+	return sig, nil
+}
+
+type builtinGroupingSig struct {
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:53]: > use `builtinGroupingImplSig`?

okk

File: expression/builtin_grouping.go (hunk 9 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:169]: > maybe getMetaGroupingID outside the loop is better.

okk

File: expression/builtin_regexp_util.go (hunk 11 above)
Comment (xzhangxian1008) [reply expression/builtin_regexp_util.go:75]: > seems this function is not used?

I will delete it.

File: parser/ast/functions.go (hunk 10 above)
Comment (xzhangxian1008) [reply parser/ast/functions.go:303]: > shoul not add this since we actually not support it yet.

I will delete it.

File: expression/builtin_grouping.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func (b *builtinGroupingSig) getMetaVersion() uint32 {
+	return b.version
+}
+
+// metadata returns the metadata of grouping functions
+func (b *builtinGroupingSig) metadata() proto.Message {
+	args := &tipb.GroupingFunctionMetadata{
+		// TODO
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:76]: > Why leave a todo here

I will implement this.

File: expression/builtin_grouping.go [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+	bf.tp.SetFlen(1)
+	sig := &builtinGroupingSig{bf, 0, map[int64]struct{}{}}
+	sig.setPbCode(tipb.ScalarFuncSig_GroupingSig)
+	return sig, nil
+}
+
+type builtinGroupingSig struct {
+	baseBuiltinFunc
+
+	// TODO these are two temporary fields for tests
```
[the remark this answers is not in the document]
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:56]: > Why the two fields are temporary?

Because I can't sure how optimizer sets these two fields, so I set them as temporary field to help my tests.

File: parser/ast/functions.go (hunk 10 above)
Comment (xzhangxian1008) [reply parser/ast/functions.go:303]: > shoul not add this since we actually not support it yet.

UT need this field, may be we have to keep it.

File: expression/builtin.go [hunk 16]
```diff
@@ -799,6 +801,7 @@ var funcs = map[string]functionClass{
 	ast.UUIDToBin:       &uuidToBinFunctionClass{baseFunctionClass{ast.UUIDToBin, 1, 2}},
 	ast.BinToUUID:       &binToUUIDFunctionClass{baseFunctionClass{ast.BinToUUID, 1, 2}},
 	ast.TiDBShard:       &tidbShardFunctionClass{baseFunctionClass{ast.TiDBShard, 1, 1}},
+	groupingImplName:    &groupingImplFunctionClass{baseFunctionClass{groupingImplName, 1, 1}},
```
Comment (windtalker) [inline expression/builtin.go:804]: Is this change still needed for this pr?

File: expression/builtin_grouping.go [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+func (b *builtinGroupingImplSig) SetMetaGroupingMarks(groupingMarks map[int64]struct{}) {
+	b.groupingMarks = groupingMarks
+}
+
+func (b *builtinGroupingImplSig) getGroupingMode() tipb.GroupingMode {
+	return b.mode
+}
+
+// metadata returns the metadata of grouping functions
+func (b *builtinGroupingImplSig) metadata() proto.Message {
```
Comment (windtalker) [inline expression/builtin_grouping.go:79]: Better to checkMetadata here.

File: expression/builtin_grouping.go [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+	args := &tipb.GroupingFunctionMetadata{}
+	*(args.Mode) = uint32(b.mode)
```
Comment (windtalker) [inline expression/builtin_grouping.go:81]: Why need to cast to `uint32`, I think both `arg.Mode` and `b.mode` is `tipb.GroupingMode`

File: expression/builtin_grouping.go [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingImplSig) checkMetadata() error {
+	mode := b.getGroupingMode()
+	grouping_ids := b.getMetaGroupingMarks()
+	if mode != tipb.GroupingMode_ModeBitAnd && mode != tipb.GroupingMode_ModeNumericCmp && mode != tipb.GroupingMode_ModeNumericSet {
+		return errors.Errorf("Mode of meta data in grouping function is invalid. input mode: %d", mode)
+	} else if (mode == 1 || mode == 2) && len(grouping_ids) != 1 {
```
Comment (windtalker) [inline expression/builtin_grouping.go:114]: why still use `mode == 1 || mode == 2` here

File: expression/builtin_grouping.go [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+		return b.groupingImplNumericCmp(groupingID, b.getMetaGroupingID())
+	case tipb.GroupingMode_ModeNumericSet:
+		return b.groupingImplNumericSet(groupingID)
+	}
+	return 0
+}
+
+// evalInt evals a builtinGroupingSig.
+func (b *builtinGroupingImplSig) evalInt(row chunk.Row) (int64, bool, error) {
+	err := b.checkMetadata()
```
Comment (windtalker) [inline expression/builtin_grouping.go:157]: It looks like very inefficient to call `checkMetadata` for every row, how about this:
1. merge `SetGroupingMode` and `SetMetaGroupingMarks` into one function like `SetMetadata`, and only call `checkMetadata` once in `SetMetadata`
2. add a flag `metaInited`, and set it to true in `SetMetadata`
3. only check `metaInited` during runtime

File: expression/builtin_grouping.go [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+			resContainer[i] = b.groupingImplNumericCmp(groupingIds.GetInt64(i), metaGroupingID)
+		}
+	case tipb.GroupingMode_ModeNumericSet:
+		for i := 0; i < rowNum; i++ {
+			resContainer[i] = b.groupingImplNumericSet(groupingIds.GetInt64(i))
+		}
+	}
+}
+
+func (b *builtinGroupingImplSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
```
Comment (windtalker) [inline expression/builtin_grouping.go:191]: The same as `evalInt`, why not check metadata here?

File: expression/builtin.go (hunk 16 above)
Comment (xzhangxian1008) [reply expression/builtin.go:804]: > Is this change still needed for this pr?

I will delete it

File: expression/builtin_grouping.go (hunk 18 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:81]: > Why need to cast to `uint32`, I think both `arg.Mode` and `b.mode` is `tipb.GroupingMode`

data type in tipb in incorrect, I will change the type in tipb first and fix it.

File: expression/builtin_grouping.go (hunk 19 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:114]: > why still use `mode == 1 || mode == 2` here

I will change it

File: expression/builtin_grouping.go (hunk 20 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:157]: > 

okk, I think it's a good idea.

File: expression/builtin_grouping.go (hunk 21 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:191]: > The same as `evalInt`, why not check metadata here?

I will check it.

File: expression/builtin_grouping.go [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+	var metaGroupingID int64
+	grouping_ids := b.getMetaGroupingMarks()
+	for key := range grouping_ids {
+		metaGroupingID = key
+	}
+	return metaGroupingID
+}
+
+func (b *builtinGroupingImplSig) checkMetadata() error {
+	mode := b.getGroupingMode()
```
Comment (windtalker) [inline expression/builtin_grouping.go:122]: should check `isMetaInited` first

**Review by windtalker — APPROVED** [review]
Others LGTM

File: expression/builtin_grouping.go (hunk 22 above)
Comment (xzhangxian1008) [reply expression/builtin_grouping.go:126]: > should check `isMetaInited` first

okk
